File: frontend/src/types/contentType.ts
export const knowledgeContentTypes = [
  'article',
  'book',
  'webpage',
  'meeting_notes',
  'report',
  'presentation',
  'spreadsheet',
  'manual',
  'other',
] as const;

export type KnowledgeContentType = typeof knowledgeContentTypes[number];

export interface ContentClassification {
  schema_version: number;
  type: KnowledgeContentType;
  source: 'ai' | 'rule' | 'manual';
  confidence?: number;
  matched_at?: string;
}

export function getContentClassification(metadata: unknown): ContentClassification | null {
  let value = metadata;
  if (typeof value === 'string') {
    try { value = JSON.parse(value); } catch { return null; }
  }
  const classification = (value as any)?.content_classification;
  return classification && knowledgeContentTypes.includes(classification.type)
    ? classification as ContentClassification
    : null;
}

export function setContentClassification(
  metadata: unknown,
  type: KnowledgeContentType,
): Record<string, unknown> {
  let value: Record<string, unknown> = {};
  if (typeof metadata === 'string') {
    try { value = JSON.parse(metadata) || {}; } catch { value = {}; }
  } else if (metadata && typeof metadata === 'object') {
    value = { ...(metadata as Record<string, unknown>) };
  }
  value.content_classification = {
    schema_version: 1,
    type,
    source: 'manual',
    confidence: 1,
    matched_at: new Date().toISOString(),
  };
  return value;
}

export function contentTypeOptions(t: (key: string) => string) {
  return knowledgeContentTypes.map((value) => ({
    value,
    label: t(`knowledgeBase.contentTypes.${value}`),
  }));
}

File: internal/types/content_type.go
package types

import (
	"encoding/json"
	"fmt"
	"time"
)

type KnowledgeContentType string

const (
	KnowledgeContentTypeArticle      KnowledgeContentType = "article"
	KnowledgeContentTypeBook         KnowledgeContentType = "book"
	KnowledgeContentTypeWebpage      KnowledgeContentType = "webpage"
	KnowledgeContentTypeMeetingNotes KnowledgeContentType = "meeting_notes"
	KnowledgeContentTypeReport       KnowledgeContentType = "report"
	KnowledgeContentTypePresentation KnowledgeContentType = "presentation"
	KnowledgeContentTypeSpreadsheet  KnowledgeContentType = "spreadsheet"
	KnowledgeContentTypeManual       KnowledgeContentType = "manual"
	KnowledgeContentTypeOther        KnowledgeContentType = "other"
)

var knowledgeContentTypes = map[KnowledgeContentType]struct{}{
	KnowledgeContentTypeArticle:      {},
	KnowledgeContentTypeBook:         {},
	KnowledgeContentTypeWebpage:      {},
	KnowledgeContentTypeMeetingNotes: {},
	KnowledgeContentTypeReport:       {},
	KnowledgeContentTypePresentation: {},
	KnowledgeContentTypeSpreadsheet:  {},
	KnowledgeContentTypeManual:       {},
	KnowledgeContentTypeOther:        {},
}

func (t KnowledgeContentType) IsValid() bool {
	_, ok := knowledgeContentTypes[t]
	return ok
}

type ContentClassificationMetadata struct {
	SchemaVersion int                  `json:"schema_version"`
	Type          KnowledgeContentType `json:"type"`
	Source        string               `json:"source"`
	Confidence    float64              `json:"confidence,omitempty"`
	MatchedAt     time.Time            `json:"matched_at"`
}

func (k *Knowledge) SetContentClassification(classification ContentClassificationMetadata) error {
	if !classification.Type.IsValid() {
		return fmt.Errorf("unsupported content type %q", classification.Type)
	}
	raw, err := k.Metadata.Map()
	if err != nil || raw == nil {
		raw = map[string]interface{}{}
	}
	raw["content_classification"] = classification
	encoded, err := json.Marshal(raw)
	if err != nil {
		return err
	}
	k.Metadata = JSON(encoded)
	return nil
}

func (k *Knowledge) ContentClassification() (*ContentClassificationMetadata, error) {
	raw, err := k.Metadata.Map()
	if err != nil || raw == nil {
		return nil, err
	}
	value, ok := raw["content_classification"]
	if !ok {
		return nil, nil
	}
	encoded, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	var classification ContentClassificationMetadata
	if err := json.Unmarshal(encoded, &classification); err != nil {
		return nil, err
	}
	return &classification, nil
}
